### PR TITLE
[rllib] Basic Offline Data IO API

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This is the first step for implementing a offline data I/O API:

What works:
```
# Generate experiences
rllib train --env=CartPole-v0 --run=PG --config='{"output": "s3://bucket/data"}'

# Consume experiences
rllib train --env=CartPole-v0 --run=DQN \
  --config='{"input": "/tmp/data", "input_evaluation": "simulation"}'

# Consume experiences from S3 files
rllib train --env=CartPole-v0 --run=DQN \
  --config='{"input": ["s3://bucket/f1", "s3://bucket/f2"], \
                  "input_evaluation": "simulation"}'

# Consume experiences from different sources
rllib train --env=CartPole-v0 --run=DQN \
  --config='{"input": {"sampler": 0.5, "/tmp/data": 0.5}}'

```

What's next:
- trajectory postprocessing and multi-agent support
- add a non-batched input format
- counterfactual evaluation (also need some way to get the reward timeline)
- documentation

## Related issue number

https://github.com/ray-project/ray/issues/3363

### CODE REVIEW

**Comment by ericl** [comment]
Cc @joneswong for comments

**Comment by ericl** [comment]
jenkins retest this please

**Comment by ericl** [comment]
Since there were a bunch of relative import issues with the 'io' package, I've renamed it to 'offline'.

**Review by richardliaw — COMMENTED** [review]
documentation, examples will come in a later PR?

File: python/ray/rllib/offline/json_reader.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+class JsonReader(InputReader):
+    """Reader object that loads experiences from JSON file chunks.
+
+    The input files will be read from in an random order."""
+
+    def __init__(self, ioctx, inputs):
+        """Initialize a JsonReader.
+
+        Arguments:
+            ioctx (IOContext): current IO context object.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/offline/json_reader.py]: This is just for consistency for all IO classes.

**Comment by ericl** [comment]
Yes, we have a lot more to implement before this is usable.

**Review by richardliaw — APPROVED** [review]

**Comment by ericl** [comment]
RLlib tests look fine.

**Pull request merged**